wp/lib/bap_wp/src/precondition.ml

```diff
@@ -314,6 +315,60 @@ let increment_stack_ptr (post : Constr.t) (env : Env.t) (offset : Exp.t)
   let z3_off, _, _, env = exp_to_z3 offset env in
   Constr.substitute_one post sp z3_off, env
 
+let get_vars (t : Sub.t) : Var.Set.t =
+  let visitor =
+    (object inherit [Var.Set.t] Term.visitor
+      method! visit_def def vars =
+        let vars = Var.Set.add vars (Def.lhs def) in
+        let vars = Var.Set.union vars (Def.free_vars def) in
+        vars
+      method! visit_jmp jmp vars =
+        Var.Set.union vars (Jmp.free_vars jmp)
+    end)
+  in
+  visitor#visit_sub t Var.Set.empty
+
+(* We are currently filtering out mem from the input variables because it causes
+   Z3 to slow down during evaluation. *)
+let get_function_inputs (sub : Sub.t) (env : Env.t) : Var.t list =
+  let arch = Env.get_arch env in
+  let module Target = (val target_of_arch arch) in
+  sub
+  |> Sub.ssa
+  |> Sub.free_vars
+  |> Var.Set.filter ~f:(fun v -> not (Target.CPU.is_mem v))
+  |> Var.Set.to_list
+
+(* Creates a Z3 function of the form func_ret_out_var(in_vars, ...) which represents
+   an output variable to a function call. It is added to the environment, and
+   substitutes the original z3_expr representing the output variable. *)
+let subst_fun_outputs (env : Env.t) (sub : Sub.t) (post : Constr.t)
```

Comment:
If you're building these here each time, why do they need to be in the `Env`?

Reply:
You're right. We don't really need them in the `env`. It would probably be better to remove that field.